compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs

```diff
@@ -48,30 +48,42 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     /// Produces warning on the given node, if the current point in the
     /// function is unreachable, and there hasn't been another warning.
     pub(crate) fn warn_if_unreachable(&self, id: HirId, span: Span, kind: &str) {
```

Comment:
I rewrote this method to split up the 3 special cases. No functional changes other than:

```
orig_span.is_desugaring(DesugaringKind::Await)
```

to 

```
span.is_desugaring(DesugaringKind::Await)
```

Since the former suppresses any unreachable warnings originating *from* an await, but we actually want to suppress any unreachable warnings that occur *within* an await.